deps/rabbit/src/lqueue.erl

```diff
@@ -7,96 +7,402 @@
-                    {Result, {L-1, Q1}}.
-
--spec out_r(?MODULE(T)) -> {result(T), ?MODULE(T)}.
-
-out_r({0, _Q} = Q) -> {empty, Q};
-out_r({L,  Q})     -> {Result, Q1} = ?QUEUE:out_r(Q),
-                      {Result, {L-1, Q1}}.
-
--spec join(?MODULE(A), ?MODULE(B)) -> ?MODULE(A | B).
-
-join({L1, Q1}, {L2, Q2}) -> {L1 + L2, ?QUEUE:join(Q1, Q2)}.
-
--spec to_list(?MODULE(T)) -> [T].
-
-to_list({_L, Q}) -> ?QUEUE:to_list(Q).
-
--spec from_list([T]) -> ?MODULE(T).
-
-from_list(L) -> {length(L), ?QUEUE:from_list(L)}.
-
--spec foldl(fun ((T, B) -> B), B, ?MODULE(T)) -> B.
-
-foldl(Fun, Init, Q) ->
-    case out(Q) of
-        {empty, _Q}      -> Init;
-        {{value, V}, Q1} -> foldl(Fun, Fun(V, Init), Q1)
-    end.
-
--spec foldr(fun ((T, B) -> B), B, ?MODULE(T)) -> B.
-
-foldr(Fun, Init, Q) ->
-    case out_r(Q) of
-        {empty, _Q}      -> Init;
-        {{value, V}, Q1} -> foldr(Fun, Fun(V, Init), Q1)
-    end.
-
--spec len(?MODULE(_)) -> non_neg_integer().
-
-len({L, _}) -> L.
-
--spec peek(?MODULE(T)) -> result(T).
-
-peek({ 0, _Q}) -> empty;
-peek({_L,  Q}) -> ?QUEUE:peek(Q).
-
--spec peek_r(?MODULE(T)) -> result(T).
-
-peek_r({ 0, _Q}) -> empty;
-peek_r({_L,  Q}) -> ?QUEUE:peek_r(Q).
+%%--------------------------------------------------------------------------
+%% Efficient implementation of double ended fifo queues
+%%
+%% Queue representation
+%%
+%% {Length,RearList,FrontList}
+%%
+%% The first element in the queue is at the head of the FrontList
+%% The last element in the queue is at the head of the RearList,
+%% that is; the RearList is reversed.
+
+%% For backwards compatibility, lqueue functions accept old state, but return only new state.
+%% Compared to the deprecated state, the new state creates less garbage per queue operation:
+%% 4 = erts_debug:size({0, [], []}).
+%% 6 = erts_debug:size({0, {[], []}}).
+-define(STATE(T),{
+                Length :: non_neg_integer(),
+                Rear :: list(T), Front :: list(T)
+               }).
+-define(DEPRECATED_STATE(T),{
+                           Length :: non_neg_integer(),
+                           %% queue:queue(T)}
+                           {Rear :: list(T), Front :: list(T)}
+                          }).
+-type ?MODULE() :: ?MODULE(_).
+-opaque ?MODULE(T) :: ?STATE(T) | ?DEPRECATED_STATE(T).
+
+%% Creation, inspection and conversion
+
+%% O(1)
+-spec new() -> ?MODULE().
+new() -> {0, [], []}.
+
+%% O(1)
+-spec is_empty(Q :: ?MODULE()) -> boolean().
+is_empty({0, [], []}) ->
+    true;
+is_empty({L, R, F})
+  when is_integer(L), L > 0, is_list(R), is_list(F) ->
+    false;
+%% accept deprecated state
+is_empty({L, {R, F}})
+  when is_integer(L), L >= 0, is_list(R), is_list(F) ->
+    is_empty({L, R, F});
+is_empty(Q) ->
+    erlang:error(badarg, [Q]).
+
+%% O(1)
+-spec len(Q :: ?MODULE()) -> non_neg_integer().
+len({L, R, F})
+  when is_integer(L), L >= 0, is_list(R), is_list(F) ->
+    L;
+%% accept deprecated state
+len({L, _Q})
+  when is_integer(L), L >= 0 ->
```

Comment:
Strictly speaking this could guard that `_Q` is actually a tuple of two lists, as elsewhere, like `to_list({L, {R, F}})
   when is_integer(L), L >= 0, is_list(R), is_list(F)`

Reply:
Thank you. I pushed a commit which adds the guard.